panic: runtime error: invalid memory address or nil pointer dereference

**What happened**:
Trying to run the example Deployment/Test as mentioned here https://kuttl.dev/docs/kuttl-test-harness.html#writing-your-first-test

This test passes as expected but if i change the number of replicas in either the Deployment or the Assert then i receive the following errors:
```
❯ kubectl kuttl test ./e2e/
=== RUN   kuttl
    kuttl: harness.go:382: starting setup
    kuttl: harness.go:229: running tests using configured kubeconfig.
    kuttl: harness.go:317: running tests
    kuttl: harness.go:70: going to run test suite with timeout of 30 seconds for each step
=== RUN   kuttl/harness
=== RUN   kuttl/harness/example
=== PAUSE kuttl/harness/example
=== CONT  kuttl/harness/example
    kuttl/harness/example: logger.go:42: 19:49:48 | example | Creating namespace: kudo-test-certain-snail
    kuttl/harness/example: logger.go:42: 19:49:48 | example/0-install | starting test step 0-install
    kuttl/harness/example: logger.go:42: 19:49:49 | example/0-install | Deployment:kudo-test-certain-snail/example-deployment created
    kuttl/harness/example: logger.go:42: 19:50:21 | example/0-install | test step failed 0-install
    kuttl/harness/example: logger.go:42: 19:50:21 | example | Deleting namespace: kudo-test-certain-snail
--- FAIL: kuttl (37.53s)
    --- FAIL: kuttl/harness (33.81s)
        --- FAIL: kuttl/harness/example (33.81s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x2028e1d]

goroutine 399 [running]:
testing.tRunner.func1.1(0x2166740, 0x3151610)
        /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc000360480)
        /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:943 +0x3f9
panic(0x2166740, 0x3151610)
        /usr/local/Cellar/go/1.14.4/libexec/src/runtime/panic.go:969 +0x166
github.com/kudobuilder/kuttl/pkg/test.(*Step).Run(0xc0000f3550, 0xc000694880, 0x17, 0xc0000cb820, 0xc000360480, 0xc0008ec8a0)
        /Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/step.go:413 +0x72d
github.com/kudobuilder/kuttl/pkg/test.(*Case).Run(0xc00038de80, 0xc000360480, 0xc000491c00)
        /Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/case.go:191 +0x393
github.com/kudobuilder/kuttl/pkg/test.(*Harness).RunTests.func1.1(0xc000360480)
        /Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/harness.go:349 +0x1da
testing.tRunner(0xc000360480, 0xc0000ca420)
        /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
        /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:1042 +0x357
```

**What you expected to happen**:
The test should run and Fail (but not error).

**How to reproduce it (as minimally and precisely as possible)**:
Create the following directory structure:
./e2e/example/

Inside the example folder create the following 2 files...
1.  Create 00-install.yaml:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: example-deployment
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
```

2 . Create 00-assert-yaml (note readyReplicas: 4, so test should fail)
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: example-deployment
status:
  readyReplicas: 4
```

Run test against a cluster (i'm using an existing AWS EKS cluster)
`kubectl kuttl test ./e2e/`

**Anything else we need to know?**:
The test Passes and doesn't error if the readyReplicas matches the deployment replicase value.

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-27T00:36:48Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.8-eks-e16311", GitCommit:"e163110a04dcb2f39c3325af96d019b4925419eb", GitTreeState:"clean", BuildDate:"2020-03-27T22:37:12Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```
- KUTTL version (use `kubectl kuttl version`): 
`KUTTL Version: version.Info{GitVersion:"0.5.0", GitCommit:"914c2ca", BuildDate:"2020-06-29T19:14:19Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}`
- Cloud provider or hardware configuration: AWS EKS
- OS (e.g. from /etc/os-release): Mac
- Kernel (e.g. `uname -a`): Darwin MC02XKA0QJG5J 18.7.0 Darwin Kernel Version 18.7.0: Mon Apr 27 20:09:39 PDT 2020; root:xnu-4903.278.35~1/RELEASE_X86_64 x86_64

--- a/pkg/test/step.go
+++ b/pkg/test/step.go
@@ -410,6 +410,9 @@ func (s *Step) Run(namespace string) []error {
 	}
 	// test failure processing
 	s.Logger.Log("test step failed", s.String())
+	if s.Assert == nil {
+		return testErrors
+	}
 	for _, collector := range s.Assert.Collectors {
 		s.Logger.Logf("collecting log output for %s", collector.String())
 		_, err := testutils.RunCommand(context.TODO(), namespace, *collector.Command(), s.Dir, s.Logger, s.Logger, s.Logger, s.Timeout)

--- a/pkg/test/step_integration_test.go
+++ b/pkg/test/step_integration_test.go
@@ -379,3 +379,54 @@ func TestTwoTestStepping(t *testing.T) {
 	err = step.LoadYAML("step_integration_test_data/two_step/01-step1.yaml")
 	assert.Error(t, err, "more than 1 TestStep not allowed in step \"twostepping\"")
 }
+
+// intentional testing that a test failure captures the test errors and does not have a segfault
+// driving by issue: https://github.com/kudobuilder/kuttl/issues/154
+func TestStepFailure(t *testing.T) {
+	// an assert without setup
+	var expected runtime.Object = &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "v1",
+			"kind":       "Pod",
+			"metadata": map[string]interface{}{
+				"labels": map[string]interface{}{
+					"app": "nginx",
+				},
+			},
+			"spec": map[string]interface{}{
+				"containers": []interface{}{
+					map[string]interface{}{
+						"image": "nginx:1.7.9",
+						"name":  "nginx",
+					},
+				},
+			},
+		},
+	}
+
+	namespace := fmt.Sprintf("kudo-test-%s", petname.Generate(2, "-"))
+
+	err := testenv.Client.Create(context.TODO(), testutils.NewResource("v1", "Namespace", namespace, ""))
+	if !k8serrors.IsAlreadyExists(err) {
+		// we are ignoring already exists here because in tests we by default use retry client so this can happen
+		assert.Nil(t, err)
+	}
+
+	//for _, actual := range test.actual {
+	//	_, _, err := testutils.Namespaced(testenv.DiscoveryClient, actual, namespace)
+	//	assert.Nil(t, err)
+	//
+	//	assert.Nil(t, testenv.Client.Create(context.TODO(), actual))
+	//}
+	asserts := []runtime.Object{expected}
+	step := Step{
+		Logger:          testutils.NewTestLogger(t, ""),
+		Client:          func(bool) (client.Client, error) { return testenv.Client, nil },
+		DiscoveryClient: func() (discovery.DiscoveryInterface, error) { return testenv.DiscoveryClient, nil },
+		Asserts:         asserts,
+		Timeout:         1,
+	}
+
+	errs := step.Run(namespace)
+	assert.Equal(t, len(errs), 1)
+}

